File: lumen_mobile/src/services/lifePlan.ts
/**
 * Life Plan API Service
 * ======================
 * Cliente para o módulo Projeto de Vida.
 * Dados sensíveis — não logar conteúdo de respostas.
 */

import api from '@/services/api';

// ── Types ──────────────────────────────────────────────────────────────────

export interface ActionOut {
  id: string;
  goal_id: string;
  action: string;
  frequency: string | null;
  context: string | null;
}

export interface GoalOut {
  id: string;
  cycle_id: string;
  is_primary: boolean;
  title: string;
  description: string | null;
  display_order: number;
  actions: ActionOut[];
}

export interface DiagnosisOut {
  id: string;
  cycle_id: string;
  dimension: string;
  abandonar: string | null;
  melhorar: string | null;
  deus_pede: string | null;
}

export interface CoreOut {
  id: string;
  cycle_id: string;
  dominant_defect: string | null;
  virtudes: string | null;
  spiritual_director_name: string | null;
  other_devotions: string | null;
}

export interface SpiritualRoutineOut {
  id: string;
  cycle_id: string;
  prayer_type: string | null;
  prayer_duration: string | null;
  mass_frequency: string | null;
  confession_frequency: string | null;
  exam_of_conscience: boolean | null;
  exam_time: string | null;
  spiritual_reading: string | null;
  spiritual_direction_frequency: string | null;
  other_practices: string | null;
}

export interface MonthlyReviewOut {
  id: string;
  cycle_id: string;
  review_date: string;
  progress_reflection: string | null;
  difficulties: string | null;
  constancy_reflection: string | null;
  decision: string;
  notes: string | null;
  created_at: string;
}

export interface CycleOut {
  id: string;
  status: string;
  realidade_vocacional: string | null;
  wizard_progress: Record<string, unknown> | null;
  started_at: string | null;
  ended_at: string | null;
  created_at: string;
  updated_at: string;
  diagnoses: DiagnosisOut[];
  core: CoreOut | null;
  goals: GoalOut[];
  routine: SpiritualRoutineOut | null;
  monthly_reviews: MonthlyReviewOut[];
}

export interface CycleSummaryOut {
  id: string;
  status: string;
  realidade_vocacional: string | null;
  started_at: string | null;
  ended_at: string | null;
  primary_goal_title: string | null;
  dominant_defect: string | null;
  review_count: number;
  created_at: string;
}

// ── Input types ────────────────────────────────────────────────────────────

export interface DiagnosisUpsert {
  dimension: string;
  abandonar?: string | null;
  melhorar?: string | null;
  deus_pede?: string | null;
}

export interface CoreUpsert {
  dominant_defect?: string | null;
  virtudes?: string | null;
  spiritual_director_name?: string | null;
  other_devotions?: string | null;
}

export interface GoalCreate {
  is_primary: boolean;
  title: string;
  description?: string | null;
  display_order?: number;
  actions?: ActionCreate[];
}

export interface GoalUpdate {
  title?: string;
  description?: string | null;
  display_order?: number;
  is_primary?: boolean;
}

export interface ActionCreate {
  action: string;
  frequency?: string | null;
  context?: string | null;
}

export interface ActionUpdate {
  action?: string;
  frequency?: string | null;
  context?: string | null;
}

export interface SpiritualRoutineUpsert {
  prayer_type?: string | null;
  prayer_duration?: string | null;
  mass_frequency?: string | null;
  confession_frequency?: string | null;
  exam_of_conscience?: boolean | null;
  exam_time?: string | null;
  spiritual_reading?: string | null;
  spiritual_direction_frequency?: string | null;
  other_practices?: string | null;
}

export interface MonthlyReviewCreate {
  review_date: string;
  progress_reflection?: string | null;
  difficulties?: string | null;
  constancy_reflection?: string | null;
  decision: string;
  notes?: string | null;
  updated_goal_id?: string | null;
  updated_goal_title?: string | null;
  updated_goal_description?: string | null;
}

// ── API functions ──────────────────────────────────────────────────────────

export const lifePlanApi = {
  // Cycles
  getActiveCycle: () =>
    api.get<CycleOut | null>('/life-plan/me/active'),

  createCycle: (data: { realidade_vocacional?: string | null }) =>
    api.post<CycleOut>('/life-plan/cycles', data as Record<string, unknown>),

  getCycle: (cycleId: string) =>
    api.get<CycleOut>(`/life-plan/cycles/${cycleId}`),

  updateWizardProgress: (cycleId: string, wizardProgress: Record<string, unknown>) =>
    api.patch<CycleOut>(`/life-plan/cycles/${cycleId}/wizard-progress`, {
      wizard_progress: wizardProgress,
    }),

  activateCycle: (cycleId: string) =>
    api.post<CycleOut>(`/life-plan/cycles/${cycleId}/activate`),

  getHistory: () =>
    api.get<CycleSummaryOut[]>('/life-plan/history'),

  // Diagnoses
  upsertDiagnosis: (cycleId: string, data: DiagnosisUpsert) =>
    api.post<DiagnosisOut>(`/life-plan/cycles/${cycleId}/diagnoses`, data as Record<string, unknown>),

  // Core
  upsertCore: (cycleId: string, data: CoreUpsert) =>
    api.post<CoreOut>(`/life-plan/cycles/${cycleId}/core`, data as Record<string, unknown>),

  // Goals
  createGoal: (cycleId: string, data: GoalCreate) =>
    api.post<GoalOut>(`/life-plan/cycles/${cycleId}/goals`, data as Record<string, unknown>),

  updateGoal: (goalId: string, data: GoalUpdate) =>
    api.patch<GoalOut>(`/life-plan/goals/${goalId}`, data as Record<string, unknown>),

  deleteGoal: (goalId: string) =>
    api.delete<void>(`/life-plan/goals/${goalId}`),

  // Actions
  createAction: (goalId: string, data: ActionCreate) =>
    api.post<ActionOut>(`/life-plan/goals/${goalId}/actions`, data as Record<string, unknown>),

  updateAction: (actionId: string, data: ActionUpdate) =>
    api.patch<ActionOut>(`/life-plan/actions/${actionId}`, data as Record<string, unknown>),

  deleteAction: (actionId: string) =>
    api.delete<void>(`/life-plan/actions/${actionId}`),

  // Routine
  upsertRoutine: (cycleId: string, data: SpiritualRoutineUpsert) =>
    api.post<SpiritualRoutineOut>(`/life-plan/cycles/${cycleId}/routine`, data as Record<string, unknown>),

  // Reviews
  createReview: (cycleId: string, data: MonthlyReviewCreate) =>
    api.post<MonthlyReviewOut>(`/life-plan/cycles/${cycleId}/reviews`, data as Record<string, unknown>),

  getReviews: (cycleId: string) =>
    api.get<MonthlyReviewOut[]>(`/life-plan/cycles/${cycleId}/reviews`),
};

export default lifePlanApi;

File: backend/app/schemas/life_plan.py
"""
Schemas — Módulo Projeto de Vida
=================================
Dados sensíveis. Não expor em logs ou relatórios de admin.
"""

from __future__ import annotations

from datetime import date, datetime
from typing import Any, Optional
from uuid import UUID

from pydantic import BaseModel, Field


# ── Diagnosis ──────────────────────────────────────────────────────────────


class DiagnosisUpsert(BaseModel):
    dimension: str  # HUMANA | ESPIRITUAL | COMUNITARIA | INTELECTUAL | APOSTOLICA
    abandonar: Optional[str] = None
    melhorar: Optional[str] = None
    deus_pede: Optional[str] = None


class DiagnosisOut(BaseModel):
    id: UUID
    cycle_id: UUID
    dimension: str
    abandonar: Optional[str]
    melhorar: Optional[str]
    deus_pede: Optional[str]

    model_config = {"from_attributes": True}


# ── Core ───────────────────────────────────────────────────────────────────


class CoreUpsert(BaseModel):
    dominant_defect: Optional[str] = None
    virtudes: Optional[str] = None
    spiritual_director_name: Optional[str] = Field(None, max_length=200)
    other_devotions: Optional[str] = None


class CoreOut(CoreUpsert):
    id: UUID
    cycle_id: UUID

    model_config = {"from_attributes": True}


# ── Action ─────────────────────────────────────────────────────────────────


class ActionCreate(BaseModel):
    action: str
    frequency: Optional[str] = Field(None, max_length=100)
    context: Optional[str] = None


class ActionUpdate(BaseModel):
    action: Optional[str] = None
    frequency: Optional[str] = Field(None, max_length=100)
    context: Optional[str] = None


class ActionOut(BaseModel):
    id: UUID
    goal_id: UUID
    action: str
    frequency: Optional[str]
    context: Optional[str]

    model_config = {"from_attributes": True}


# ── Goal ───────────────────────────────────────────────────────────────────


class GoalCreate(BaseModel):
    is_primary: bool = False
    title: str = Field(..., max_length=80)
    description: Optional[str] = None
    display_order: int = 0
    actions: list[ActionCreate] = []


class GoalUpdate(BaseModel):
    title: Optional[str] = Field(None, max_length=80)
    description: Optional[str] = None
    display_order: Optional[int] = None
    is_primary: Optional[bool] = None


class GoalOut(BaseModel):
    id: UUID
    cycle_id: UUID
    is_primary: bool
    title: str
    description: Optional[str]
    display_order: int
    actions: list[ActionOut] = []

    model_config = {"from_attributes": True}


# ── Spiritual Routine ──────────────────────────────────────────────────────


class SpiritualRoutineUpsert(BaseModel):
    prayer_type: Optional[str] = Field(None, max_length=200)
    prayer_duration: Optional[str] = Field(None, max_length=100)
    mass_frequency: Optional[str] = Field(None, max_length=20)
    confession_frequency: Optional[str] = Field(None, max_length=20)
    exam_of_conscience: Optional[bool] = None
    exam_time: Optional[str] = Field(None, max_length=100)
    spiritual_reading: Optional[str] = Field(None, max_length=200)
    spiritual_direction_frequency: Optional[str] = Field(None, max_length=100)
    other_practices: Optional[str] = None


class SpiritualRoutineOut(SpiritualRoutineUpsert):
    id: UUID
    cycle_id: UUID

    model_config = {"from_attributes": True}


# ── Monthly Review ─────────────────────────────────────────────────────────


class MonthlyReviewCreate(BaseModel):
    review_date: date
    progress_reflection: Optional[str] = None
    difficulties: Optional[str] = None
    constancy_reflection: Optional[str] = None
    decision: str  # CONTINUE | ADJUST_GOAL | CHANGE_PRIMARY_GOAL | NEW_CYCLE
    notes: Optional[str] = None
    # For ADJUST_GOAL / CHANGE_PRIMARY_GOAL
    updated_goal_id: Optional[UUID] = None
    updated_goal_title: Optional[str] = Field(None, max_length=80)
    updated_goal_description: Optional[str] = None


class MonthlyReviewOut(BaseModel):
    id: UUID
    cycle_id: UUID
    review_date: date
    progress_reflection: Optional[str]
    difficulties: Optional[str]
    constancy_reflection: Optional[str]
    decision: str
    notes: Optional[str]
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Cycle ──────────────────────────────────────────────────────────────────


class CycleCreate(BaseModel):
    realidade_vocacional: Optional[str] = None


class WizardProgressUpdate(BaseModel):
    wizard_progress: dict[str, Any]


class CycleOut(BaseModel):
    id: UUID
    status: str
    realidade_vocacional: Optional[str]
    wizard_progress: Optional[dict[str, Any]]
    started_at: Optional[date]
    ended_at: Optional[date]
    created_at: datetime
    updated_at: datetime
    diagnoses: list[DiagnosisOut] = []
    core: Optional[CoreOut] = None
    goals: list[GoalOut] = []
    routine: Optional[SpiritualRoutineOut] = None
    monthly_reviews: list[MonthlyReviewOut] = []

    model_config = {"from_attributes": True}


class CycleSummaryOut(BaseModel):
    id: UUID
    status: str
    realidade_vocacional: Optional[str]
    started_at: Optional[date]
    ended_at: Optional[date]
    primary_goal_title: Optional[str] = None
    dominant_defect: Optional[str] = None
    review_count: int = 0
    created_at: datetime

    model_config = {"from_attributes": True}
